presto-main/src/main/java/com/facebook/presto/metadata/FunctionManager.java

```diff
@@ -169,23 +171,29 @@ public void registerBuiltInFunctions(List<? extends BuiltInFunction> functions)
      * @throws PrestoException if there are no matches or multiple matches
      */
     public FunctionHandle resolveFunction(Session session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
-    {
-        return resolveFunction(Optional.of(session), name, parameterTypes);
-    }
-
-    private FunctionHandle resolveFunction(Optional<Session> session, QualifiedName name, List<TypeSignatureProvider> parameterTypes)
     {
         QualifiedFunctionName functionName;
         if (!name.getPrefix().isPresent()) {
             functionName = QualifiedFunctionName.of(DEFAULT_NAMESPACE, name.getSuffix());
         }
         else {
-            functionName = QualifiedFunctionName.of(name.getOriginalParts());
+            checkArgument(name.getOriginalParts().size() == 3, "Require 3 parts for qualified function name");
```

Comment:
This should throw a `PrestoException` with proper error message that's easy to understand for user. Something along the line of `Reference to non-builtin functions must be using qualified function name in the form of "a.c.b"` if `originalParts = 2`, and throwing `FUNCTION_NOT_FOUND` is reasonable if `originalParts > 3`. The 3 parts function name should be a side effect of how function namespaces are managed. Since all sql function names would have 3 parts, a >3 parts function can not be resolved.